### PR TITLE
Adds digit locomotion examples

### PR DESCRIPTION
**Pull request by lgulich**
# Description

Add an example to train a locomotion and loco-manipulation controller for digit. This also serves as an example on how to train a robot with closed loops.

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/__init__.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+gym.register(
+    id="Isaac-Velocity-Flat-Digit-Play-v0",
+    entry_point="isaaclab.envs:ManagerBasedRLEnv",
+    disable_env_checker=True,
+    kwargs={
+        "env_cfg_entry_point": f"{__name__}.flat_env_cfg:DigitFlatEnvCfg_PLAY",
+        "rsl_rl_cfg_entry_point": f"{agents.__name__}.rsl_rl_ppo_cfg:DigitFlatPPORunnerCfg",
+    },
+)
+
```
Comment (lgulich) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/__init__.py:33]:
```suggestion


```

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/agents/rsl_rl_ppo_cfg.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Copyright (c) 2022-2025, The Isaac Lab Project Developers.
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+# Copyright (c) 2022-2024, The IsaacSim RL Project Developers.
+# All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
```
Comment (lgulich) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/agents/rsl_rl_ppo_cfg.py]:
```suggestion
```

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/loco_manip_env_cfg.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
+# Copyright (c) 2022-2025, The Isaac Lab Project Developers.
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/loco_manip_env_cfg.py]: To me this task doesn't seem to fit well under "velocity" set of tasks. I suggest moving it outside to `locomanipulation` folder at the same level as `locomotion` and we call the task something like `tracking`. Locomotion velocity is only for walking :)

File: source/isaaclab_assets/isaaclab_assets/robots/agility.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import isaaclab.sim as sim_utils
+from isaaclab.actuators import IdealPDActuatorCfg, ImplicitActuatorCfg
+from isaaclab.assets.articulation import ArticulationCfg
+from isaaclab.utils.assets import ISAAC_NUCLEUS_DIR
+
+FILE_DIR = pathlib.Path(__file__).parent
+
+
+def get_digit_articulation_cfg(
```
Comment (Mayankm96) [inline source/isaaclab_assets/isaaclab_assets/robots/agility.py]: I would suggest to not have this functional way of asset configuration. Instead it is better to just have explicit configurations for each style (for consistency with the rest of the examples).

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/flat_env_cfg.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            },
+            "velocity_range": {
+                "x": (-0.0, 0.0),
+                "y": (-0.0, 0.0),
+                "z": (-0.0, 0.0),
+                "roll": (-0.0, 0.0),
+                "pitch": (-0.0, 0.0),
+                "yaw": (-0.0, 0.0),
+            },
+        }
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/flat_env_cfg.py]: Any particular reason to set all these to zero?

Comment (Mayankm96) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/flat_env_cfg.py]: Btw: If you leave the params as blank dictionary, it is implicitly understood as zero.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    ".*_hip_roll",
+    ".*_hip_yaw",
+    ".*_hip_pitch",
+    ".*_knee",
+    ".*_toe_a",
+    ".*_toe_b",
+]
+
+
+ARM_JOINT_NAMES = [".*_arm_.*"]
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: Might make it sense to move it to assets config of DIGIT to reuse it at other places as needed. Most asset related information should be at the same place IMO so that when we update the asset, we can change the USD and properties together.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py [hunk 7]
```diff
@@ -104,3 +104,21 @@ def track_ang_vel_z_world_exp(
...
+def stand_still_error(env, command_name: str, asset_cfg: SceneEntityCfg = SceneEntityCfg("robot")) -> torch.Tensor:
+    """Penalize offsets from the default joint positions when the command is very small."""
+    asset = env.scene[asset_cfg.name]
+    command = env.command_manager.get_command(command_name)
+    # Penalize motion at zero commands
+    offset = asset.data.default_joint_pos[:, asset_cfg.joint_ids] - asset.data.joint_pos[:, asset_cfg.joint_ids]
+    return torch.sum(torch.abs(offset), dim=1) * (torch.norm(command[:, :2], dim=1) < 0.06)
+
+
+def no_jumps(env, sensor_cfg: SceneEntityCfg) -> torch.Tensor:
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: We already have a MDP term called: undesired_contacts which does the same operation?

https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.envs.mdp.html#isaaclab.envs.mdp.rewards.undesired_contacts

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py [hunk 8]
```diff
@@ -104,3 +104,21 @@ def track_ang_vel_z_world_exp(
...
     return torch.exp(-ang_vel_error / std**2)
+
+
+def stand_still_error(env, command_name: str, asset_cfg: SceneEntityCfg = SceneEntityCfg("robot")) -> torch.Tensor:
+    """Penalize offsets from the default joint positions when the command is very small."""
+    asset = env.scene[asset_cfg.name]
+    command = env.command_manager.get_command(command_name)
+    # Penalize motion at zero commands
+    offset = asset.data.default_joint_pos[:, asset_cfg.joint_ids] - asset.data.joint_pos[:, asset_cfg.joint_ids]
+    return torch.sum(torch.abs(offset), dim=1) * (torch.norm(command[:, :2], dim=1) < 0.06)
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]:
```suggestion
    command = env.command_manager.get_command(command_name)
    # Penalize motion at zero commands
    return mdp.joint_deviation_l1(asset_cfg) * (torch.norm(command[:, :2], dim=1) < 0.06)
```

Comment (Mayankm96) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: Makes sense to make threshold for command also a argument to the function as it depends on the "size" of the robot or how sensitive you want this term to be.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py [hunk 9]
```diff
@@ -104,3 +104,21 @@ def track_ang_vel_z_world_exp(
     asset = env.scene[asset_cfg.name]
     ang_vel_error = torch.square(env.command_manager.get_command(command_name)[:, 2] - asset.data.root_ang_vel_w[:, 2])
     return torch.exp(-ang_vel_error / std**2)
+
+
+def stand_still_error(env, command_name: str, asset_cfg: SceneEntityCfg = SceneEntityCfg("robot")) -> torch.Tensor:
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: Maybe more descriptive on the type of error as there could be other type of stand-still errors like velocity or base pose.

```suggestion
def stand_still_joint_deviation_l1(env, command_name: str, asset_cfg: SceneEntityCfg = SceneEntityCfg("robot")) -> torch.Tensor:
```

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+    joint_pos = mdp.JointPositionActionCfg(
+        asset_name="robot",
+        joint_names=LEG_JOINT_NAMES + ARM_JOINT_NAMES,
+        scale=0.5,
+        use_default_offset=True,
+    )
+
+
+@configclass
+class EventCfg:
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: The events seem a duplication of `LocomotionVelocityEnvCfg` with just the renaming of base link. If that's the only/main change, I suggest not duplicating and just renaming in the inherited config.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        self.scene.terrain.max_init_terrain_level = None
+        # Reduce the number of terrains to save memory.
+        if self.scene.terrain.terrain_generator is not None:
+            self.scene.terrain.terrain_generator.num_rows = 5
+            self.scene.terrain.terrain_generator.num_cols = 5
+            self.scene.terrain.terrain_generator.curriculum = False
+
+        self.commands.base_velocity.ranges.lin_vel_x = (1.0, 1.0)
+        self.commands.base_velocity.ranges.lin_vel_y = (0.0, 0.0)
+        self.commands.base_velocity.ranges.heading = (0.0, 0.0)
```
Comment (Mayankm96) [inline source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: Why not keep it random?

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py (hunk 6 above)
Comment (Mayankm96) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: At least in this example, everywhere leg and arm are combined so don't see why you need it.

File: source/isaaclab_assets/isaaclab_assets/robots/agility.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+import pathlib
+from typing import Literal
+
+import isaaclab.sim as sim_utils
+from isaaclab.actuators import IdealPDActuatorCfg, ImplicitActuatorCfg
+from isaaclab.assets.articulation import ArticulationCfg
+from isaaclab.utils.assets import ISAAC_NUCLEUS_DIR
+
+FILE_DIR = pathlib.Path(__file__).parent
```
Comment (Mayankm96) [inline source/isaaclab_assets/isaaclab_assets/robots/agility.py]: Unused variable?

**Comment by Mayankm96** [comment]
Added a few comments. Would be good to also supplement the docs with images from each task :)

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/flat_env_cfg.py (hunk 5 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/flat_env_cfg.py]: that's a nice trick! I've removed the overwrites for now. I think they were originally there because the setup was used to record some videos for a presentation.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/loco_manip_env_cfg.py (hunk 3 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/loco_manip_env_cfg.py]: Created a new folder called `loco_manipulation`. We also call it LocoManipulation in other places so I fee it should have an underscore in snake_case.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py (hunk 6 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: Moved it to the agility.py file. 

I think it's nice to keep arms and legs separate, makes it easier if someone want to train just a lower body policy or just upper body policy.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py (hunk 10 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: good point, thanks. Removed it.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py (hunk 11 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/config/digit/rough_env_cfg.py]: I think this was originally used to record videos for a presentation. Removed it now.

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py (hunk 9 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: much better name

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py (hunk 8 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: Added the threshold

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py (hunk 7 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: isn't this the opposite? The `no_jumps` reward should encourage that at least 1 foot is on the floor. The `undesired_contacts` reward would encourage that no foot is on the floor, no?

File: source/isaaclab_assets/isaaclab_assets/robots/agility.py (hunk 12 above)
Comment (lgulich) [reply source/isaaclab_assets/isaaclab_assets/robots/agility.py]: done

File: source/isaaclab_assets/isaaclab_assets/robots/agility.py (hunk 4 above)
Comment (lgulich) [reply source/isaaclab_assets/isaaclab_assets/robots/agility.py]: For the sake of consistency I agree. But it does not help with removing duplication :disappointed: 

Anyways I changed it for now as we probably anyways don't need the config options to hang it or to change the actuator type.

**Review by lgulich — COMMENTED** [review]
Thanks for the review. Addressed your comments @Mayankm96

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py (hunk 7 above)
Comment (Mayankm96) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: Ah okay. Then I'd say we should make another term in the `mdp.rewards` called `desired_contacts`? That might be useful for some other tasks in the future too

**Comment by dementrock** [comment]
For this to serve as a more helpful example, would it be possible to also explain how the usd file was created?

File: source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py (hunk 7 above)
Comment (lgulich) [reply source/isaaclab_tasks/isaaclab_tasks/manager_based/locomotion/velocity/mdp/rewards.py]: done
